nexus/src/app/affinity.rs

```diff
@@ -0,0 +1,407 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Affinity groups
+
+use nexus_db_model::AffinityGroup;
+use nexus_db_model::AntiAffinityGroup;
+use nexus_db_queries::authz;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::lookup;
+use nexus_db_queries::db::lookup::LookupPath;
+use nexus_types::external_api::params;
+use nexus_types::external_api::views;
+use omicron_common::api::external;
+use omicron_common::api::external::http_pagination::PaginatedBy;
+use omicron_common::api::external::CreateResult;
+use omicron_common::api::external::DeleteResult;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::LookupResult;
+use omicron_common::api::external::NameOrId;
+use omicron_common::api::external::UpdateResult;
+use omicron_uuid_kinds::GenericUuid;
+use omicron_uuid_kinds::InstanceUuid;
+
+impl super::Nexus {
+    pub fn affinity_group_lookup<'a>(
+        &'a self,
+        opctx: &'a OpContext,
+        affinity_group_selector: params::AffinityGroupSelector,
+    ) -> LookupResult<lookup::AffinityGroup<'a>> {
+        match affinity_group_selector {
+            params::AffinityGroupSelector {
+                affinity_group: NameOrId::Id(id),
+                project: None
+            } => {
+                let affinity_group =
+                    LookupPath::new(opctx, &self.db_datastore).affinity_group_id(id);
+                Ok(affinity_group)
+            }
+            params::AffinityGroupSelector {
+                affinity_group: NameOrId::Name(name),
+                project: Some(project)
+            } => {
+                let affinity_group = self
+                    .project_lookup(opctx, params::ProjectSelector { project })?
+                    .affinity_group_name_owned(name.into());
+                Ok(affinity_group)
+            }
+            params::AffinityGroupSelector {
+                affinity_group: NameOrId::Id(_),
+                ..
+            } => {
+                Err(Error::invalid_request(
+                    "when providing affinity_group as an ID project should not be specified",
```

Comment:
nit: this reads like "(ID project) should not be specified", which is a bit nonsensical, so i think it should have a comma
```suggestion
                    "when providing affinity_group as an ID, project should not be specified",
```

Reply:
Yes, and unfortunately there are quite a few more: https://github.com/oxidecomputer/omicron/issues/7298